xfer/model_handler/model_handler.py

```diff
@@ -65,78 +65,120 @@ def drop_layer_bottom(self, num_layers_to_drop=1):
-        symbol_nodes = symbol_nodes[0:1] + symbol_nodes[n_layer_idx+1:]
-        id2name = {count: node[consts.LAYER_NAME] for count, node in enumerate(symbol_nodes)}
-        # Recreate model from symbol nodes
-        sym = self._model_from_nodes(None, symbol_nodes, elements_offset=-n_layer_idx, id2name=id2name,
-                                     data_name=self.data_name)
+        symbol_dict = self._get_symbol_dict(self.symbol)
+        # Determine number of nodes to delete by counting node ids which do not appear in arg_nodes
+        count = 0
+        for node_idx, _ in enumerate(symbol_dict[consts.NODES]):
+            if node_idx not in symbol_dict[consts.ARG_NODES]:
+                count += 1
+                if count >= num_layers_to_drop:
+                    break
+        delete_idx = node_idx  # We will delete nodes from 1 to delete_idx
+        # Delete nodes from symbol dict
+        for i in reversed(range(1, delete_idx+1)):
+            del symbol_dict[consts.NODES][i]
+        # Update arg_nodes (list of nodes that do not contain operators)
+        symbol_dict[consts.ARG_NODES] = [0] + [i-delete_idx
+                                               for c, i in enumerate(symbol_dict[consts.ARG_NODES]) if i > delete_idx]
+        # Updates heads - gives node idx of output
+        for j in symbol_dict[consts.HEADS]:
+            j[0] -= delete_idx
+        # Update node_row_ptr (list of ids of all nodes)
+        symbol_dict[consts.NODE_PTR] = symbol_dict[consts.NODE_PTR][:-delete_idx]
+        # Update inputs of nodes
+        for node in symbol_dict[consts.NODES]:
+            for ip in node[consts.INPUTS]:
+                ip[0] -= delete_idx
+        sym = mx.sym.load_json(json.dumps(symbol_dict))
+
         logging.info('{} deleted from model bottom'.format(', '.join(self.layer_names[:num_layers_to_drop])))
         self.update_sym(sym)
 
-    def add_layer_top(self, layer_factory_list):
+    def add_layer_top(self, layer_list):
         """
         Add layer to output of model.
 
-        :param layer_factory: List of LayerFactory objects to be added to model output.
-        :type layer_factory: list(:class:`LayerFactory`)
+        :param layer_list: List of MxNet symbol layers to be added to model output.
+        :type layer_list: list(:class:`mx.symbol`)
         """
         if '_label' in self.symbol.get_internals().list_outputs()[consts.LABEL_IDX]:
             raise exceptions.ModelError('Cannot add layer above output layer')
-        sym = self.symbol
         added_layer_names = []
-        for lf in layer_factory_list:
-            self._validate_layer_name(lf.attributes[consts.NAME])
-            sym = lf.create_layer(sym)
-            added_layer_names.append(lf.attributes[consts.NAME])
+
+        net_symbol_dict = self._get_symbol_dict(self.symbol)
+        for layer_symbol in layer_list:
+            added_layer_names.append(layer_symbol.name)
```

Comment:
Consider incrementing this at the end of the loop after `net_symbol_dict` update is complete